ddl/ddl_api.go

```diff
@@ -430,6 +430,8 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef, o
 				col.GeneratedStored = v.Stored
 				_, dependColNames := findDependedColumnNames(colDef)
 				col.Dependences = dependColNames
+			case ast.ColumnOptionCollate:
```

Comment:
I'm afraid you can't simply set the collate with out any check.
In This PR
```sql
test > create table t1(a varchar(10) collate utf8_bin, b decimal collate utf8_bin);
test > show create table t1
+-------+------------------------------------------------------------------------+
| Table | Create Table                                                           |
+-------+------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (                                                    |
|       |   `a` varchar(10) CHARACTER SET utf8mb4 COLLATE utf8_bin DEFAULT NULL, |
|       |   `b` decimal(11,0) DEFAULT NULL                                       |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin            |
+-------+------------------------------------------------------------------------+
test > show full columns from t1;
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
| Field | Type          | Collation | Null | Key | Default | Extra | Privileges                      | Comment |
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
| a     | varchar(10)   | utf8_bin  | YES  |     | <null>  |       | select,insert,update,references |         |
| b     | decimal(11,0) | utf8_bin  | YES  |     | <null>  |       | select,insert,update,references |         |
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
```

In MYSQL
```sql
test > create table t1(a varchar(10) collate utf8_bin, b decimal collate utf8_bin);
test > show create table t
(1146, u"Table 'test.t' doesn't exist")
test > show create table t1
+-------+---------------------------------------------------------------------+
| Table | Create Table                                                        |
+-------+---------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (                                                 |
|       |   `a` varchar(10) CHARACTER SET utf8 COLLATE utf8_bin DEFAULT NULL, |
|       |   `b` decimal(10,0) DEFAULT NULL                                    |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=ascii                               |
+-------+---------------------------------------------------------------------+
test > show full columns from t1
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
| Field | Type          | Collation | Null | Key | Default | Extra | Privileges                      | Comment |
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
| a     | varchar(10)   | utf8_bin  | YES  |     | <null>  |       | select,insert,update,references |         |
| b     | decimal(10,0) | <null>    | YES  |     | <null>  |       | select,insert,update,references |         |
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
```

Reply:
yeah, I was planning to put the checks in the subsequent PRs

Reply:
hi @crazycs520, the case you showed has been fixed in https://github.com/pingcap/tidb/pull/10007.

Another known issue is the correlation between Charset/Collation: https://github.com/pingcap/tidb/issues/10122, I plan to fix it in another PR to keep this one clear enough.

PTAL

Reply:
em... I prefer to add a check here: 
```go
                if field_types.HasCharset(colDef.Tp) {
					col.FieldType.Collate = v.StrValue
				}
```

Reply:
This is different from MySQL's way, but I think it's fine to do so

Reply:
addressed, PTAL @crazycs520